### PR TITLE
Add new case input_keyboard_basic to tp-qemu

### PR DESCRIPTION
**Pull request by peixiu**
Add new file input_keyboard_basic.py and input_keyboard_basic.cfg

id:1724059
Signed-off-by: Peixiu Hou <phou@redhat.com>

### CODE REVIEW

**Comment by peixiu** [comment]
This patch based to follows patch:
https://github.com/avocado-framework/avocado-vt/pull/2180
https://github.com/autotest/tp-qemu/pull/1852
https://github.com/autotest/tp-qemu/pull/1851
 
For CI check fail, due to module in patch 1852 not be merged caused.

**Comment by peixiu** [comment]
(1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.input_keyboard_basic.q35: PASS (139.59 s)

**Comment by peixiu** [comment]
(1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.input_keyboard_basic.q35: PASS (64.48 s)

**Comment by peixiu** [comment]
Tested with patch https://github.com/avocado-framework/avocado-vt/pull/2180, https://github.com/avocado-framework/avocado-vt/pull/2181, https://github.com/autotest/tp-qemu/pull/1852 and https://github.com/autotest/tp-qemu/pull/1851:

 (1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.input_keyboard_basic.q35: PASS (137.73 s)

**Comment by peixiu** [comment]
@luckyh updated the tested keys to cfg file, could please help to review it? thanks~

**Comment by peixiu** [comment]
(1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.input_keyboard_basic.q35: PASS (361.40 s)

**Comment by peixiu** [comment]
(1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.input_keyboard_basic.q35: PASS (299.77 s)

**Comment by peixiu** [comment]
(1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.input_keyboard_basic.q35: PASS (364.47 s)

**Comment by peixiu** [comment]
Tested passed:
 (1/4) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.input_keyboard_basic.q35: PASS (363.38 s)

**Comment by peixiu** [comment]
(1/4) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.input_keyboard_basic.q35: PASS (364.42 s)

**Comment by peixiu** [comment]
@luckyh Could you help to review this patch?
 I also create a  qemu/deps/keymaps/keymap_to_linux.json under qemu/deps directory, there isn't shared directory under qemu, so I set it to deps directory.

Thanks~

**Comment by peixiu** [comment]
(1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2008.x86_64.r2.io-github-autotest-qemu.vioinput_keyboard.q35: PASS (256.01 s)

**Comment by peixiu** [comment]
(1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.vioinput_keyboard: PASS (245.26 s)

**Comment by peixiu** [comment]
@luckyh Hi xuhan, I updated the code, please help to review it, thanks so much~

File: qemu/tests/cfg/vioinput_keyboard.cfg [hunk 1]
```diff
@@ -0,0 +1,11 @@
+- vioinput_keyboard:
+    only Windows
+    type = vioinput_keyboard
+    driver_name = "vioinput viohidkmdf hidclass.sys hidparse.sys"
+    timeout = 360
+    no Host_RHEL.m6 Host_RHEL.m7.u0 Host_RHEL.m7.u1 Host_RHEL.m7.u2 Host_RHEL.m7.u3
```
Comment (luckyh) [inline qemu/tests/cfg/vioinput_keyboard.cfg]: Suggest `required_qemu = [2.4.0, )` instead.

File: qemu/tests/vioinput_keyboard.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    :param test: kvm test object
+    :param params: dictionary with the test parameters
+    :param vm: vm object
+    :param key_name: tested keyname, default key_name='', will covered all keys
+                     on the keyboard(us).
+    :param hold_time: the time that keep one key pressed, default hole_time=0.
+    :param interval: the interval between two keys, default interval=0.
+    """
+
+    def get_keycode_cfg(filename):
```
Comment (luckyh) [inline qemu/tests/vioinput_keyboard.py]: please move it to the top level of this module.

File: qemu/tests/vioinput_keyboard.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        else:
+            key_lst = [key_check_cfg[key]]
+        key_num = len(key_lst)
+        key_event_lst = list()
+
+        while not events_queue.empty():
+            events = events_queue.get()
+            key_event_lst.append((events["keyCode"], events["type"]))
+
+        if len(key_event_lst) < 2 * key_num:
```
Comment (luckyh) [inline qemu/tests/vioinput_keyboard.py]: Doesn't it only work for `hold_time == 0` as well?

File: qemu/tests/vioinput_keyboard.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+            test.fail("Reveived key events %s were not enough" % key_event_lst)
+
+        key_up_lst = list()
+        for k, v in key_event_lst[-key_num:]:
```
Comment (luckyh) [inline qemu/tests/vioinput_keyboard.py]: Why don't you just go iterating the list forward, and check every event one by one?

File: qemu/tests/vioinput_keyboard.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    key_check_cfg = get_keycode_cfg(key_table_file)
+    keys = [key_name] if key_name else key_check_cfg.keys()
+    wait_time = params.get("wait_time", 2)
+
+    error_context.context("Start event listener in guest", logging.info)
+    listener = input_event_proxy.EventListener(vm)
+
+    error_context.context("Send key and check received event is correct",
+                          logging.info)
+    keyboard = graphical_console.GraphicalConsole(vm)
```
Comment (luckyh) [inline qemu/tests/vioinput_keyboard.py]: `console = ...`

File: qemu/tests/cfg/vioinput_keyboard.cfg [hunk 6]
```diff
@@ -0,0 +1,11 @@
+- vioinput_keyboard:
+    only Windows
+    type = vioinput_keyboard
+    driver_name = "vioinput viohidkmdf hidclass.sys hidparse.sys"
+    timeout = 360
```
Comment (luckyh) [inline qemu/tests/cfg/vioinput_keyboard.cfg]: unused param

Comment (peixiu) [reply qemu/tests/cfg/vioinput_keyboard.cfg]: thanks, will update~

File: qemu/tests/cfg/vioinput_keyboard.cfg (hunk 1 above)
Comment (peixiu) [reply qemu/tests/cfg/vioinput_keyboard.cfg]: thanks, will update~

File: qemu/tests/vioinput_keyboard.py (hunk 2 above)
Comment (peixiu) [reply qemu/tests/vioinput_keyboard.py]: ok, thanks, will update~

File: qemu/tests/vioinput_keyboard.py (hunk 3 above)
Comment (peixiu) [reply qemu/tests/vioinput_keyboard.py]: yes, it work for hold_time=0 or hold_time>0, any way, key_event_list have 2*key_num values at least.

File: qemu/tests/vioinput_keyboard.py (hunk 4 above)
Comment (peixiu) [reply qemu/tests/vioinput_keyboard.py]: Hi xuhan, for interval=0, received combination keys may are disordered, I want separate the key-up events and key-down events here, check key status firstly, then compare their keycodes with expected keys directly, thus I also can ignore order conveniently,  I just found this way, if have better one, I can optimize it, thanks~

File: qemu/tests/vioinput_keyboard.py (hunk 5 above)
Comment (peixiu) [reply qemu/tests/vioinput_keyboard.py]: that's ok, will update~

**Comment by peixiu** [comment]
@luckyh Hi xuhan, I updated the codes, could help review it again? thanks a lot~

**Comment by peixiu** [comment]
(1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.vioinput_keyboard: PASS (46.11 s)

**Pull request merged**